Add Log enrinchment IP address to ASN, City, and Country

hitenkoku opened the pull request:
## What Changed
- Added --geo-ip option in csv-timeline and json-timeline

## Evidence

### Environment
- OS: Windows 11
- Hard: Memory 16GB , Core 8, SSD, 

### Benchmark1
I ran a benchmark using [this procedure(6.1GB evtx)](https://github.com/Yamato-Security/hayabusa/issues/778#issuecomment-1296504766) and the results were as follows.

|Version| Elapsed time | Memory(peak)  | Events with hits / Total events | Output file size(bytes)|
|-------| ------------ | --------|--------|--------| 
| [before](https://github.com/Yamato-Security/hayabusa/commit/b97b22403eb4ef4633b7e5a32b499051ae7e395e)  | 00:03:52.226| 4.8 GiB | 1,593,715 / 4,817,181 | 575078117 |
| [This PR](https://github.com/Yamato-Security/hayabusa/pull/896/commits/2d89759ab5e56c16a69e6ba7b8fc6641377b5783)   |  00:03:59.750| 5.7 GiB | 1,593,715 / 4,817,181 | 614236764 | 

#### Console output

[before](https://github.com/Yamato-Security/hayabusa/tree/imporve-speed-by-using-starts-ends-with) 
```
>./main.exe csv-timeline -d ..\all-evtx\ -q --debug -o main.csv
...
Analyzing event files: 1858
Total file size: 6.1 GB

Loading detections rules. Please wait.

Excluded rules: 15
Noisy rules: 7 (Disabled)

Experimental rules: 1916 (56.59%)
Stable rules: 224 (6.62%)
Test rules: 1246 (36.80%)

Hayabusa rules: 148
Sigma rules: 3238
Total enabled detection rules: 3386

...
Results Summary:

Events with hits / Total events: 1,593,715 / 4,817,181 (Data reduction: 3,223,466 events (66.92%))

Total | Unique detections: 1,627,283 | 150
Total | Unique critical detections: 0 (0.00%) | 0 (0.00%)
Total | Unique high detections: 12,044 (0.74%) | 20 (13.33%)
Total | Unique medium detections: 11,118 (0.68%) | 38 (25.33%)
Total | Unique low detections: 1,053,622 (64.75%) | 42 (28.00%)
Total | Unique informational detections: 550,499 (33.83%) | 50 (33.33%)

...

Saved file: main.csv (575.1 MB)
Elapsed time: 00:03:52.226
Rule Parse Processing Time: 00:00:02.760
Analysis Processing Time: 00:03:38.624
Output Processing Time: 00:00:10.841

Memory usage stats:
heap stats:    peak      total      freed    current       unit      count
  reserved:    4.8 GiB    4.8 GiB   83.0 MiB    4.8 GiB
 committed:    4.1 GiB   63.0 GiB   59.0 GiB    4.0 GiB
     reset:      0          0          0          0                            ok
   touched:   64.2 KiB   35.0 MiB   56.6 GiB  -56.6 GiB                        ok
  segments:     30        561        554          7                            not all freed!
-abandoned:      0          0          0          0                            ok
   -cached:      0          0          0          0                            ok
     pages:      0          0      743.7 Ki  -743.7 Ki                         ok
-abandoned:      0          0          0          0                            ok
 -extended:      0
 -noretire:      0
     mmaps:      0
   commits:   54.2 Ki
   threads:     32         32          0         32                            not all freed!
  searches:     0.0 avg
numa nodes:       1
   elapsed:     232.235 s
   process: user: 2287.359 s, system: 16.812 s, faults: 5408453, rss: 4.0 GiB, commit: 4.1 GiB
```

[This PR](https://github.com/Yamato-Security/hayabusa/tree/879-new-feature-log-enrinchment-ip-address-to-asn-city-country)
```
>./879.exe csv-timeline -d ..\all-evtx\ -q --geo-ip ..\mmdb --debug -o 879.csv
...
Analyzing event files: 1858
Total file size: 6.1 GB

Loading detections rules. Please wait.

Excluded rules: 15
Noisy rules: 7 (Disabled)

Experimental rules: 1916 (56.59%)
Stable rules: 224 (6.62%)
Test rules: 1246 (36.80%)

Hayabusa rules: 148
Sigma rules: 3238
Total enabled detection rules: 3386

...
Results Summary:

Events with hits / Total events: 1,593,715 / 4,817,181 (Data reduction: 3,223,466 events (66.92%))

Total | Unique detections: 1,627,283 | 150
Total | Unique critical detections: 0 (0.00%) | 0 (0.00%)
Total | Unique high detections: 12,044 (0.74%) | 20 (13.33%)
Total | Unique medium detections: 11,118 (0.68%) | 38 (25.33%)
Total | Unique low detections: 1,053,622 (64.75%) | 42 (28.00%)
Total | Unique informational detections: 550,499 (33.83%) | 50 (33.33%)

...

Saved file: 879.csv (614.2 MB)
Elapsed time: 00:03:59.750
Rule Parse Processing Time: 00:00:01.517
Analysis Processing Time: 00:03:46.674
Output Processing Time: 00:00:11.558

Memory usage stats:
heap stats:    peak      total      freed    current       unit      count
  reserved:    5.7 GiB    5.7 GiB   83.0 MiB    5.6 GiB
 committed:    4.9 GiB   64.8 GiB   60.0 GiB    4.8 GiB
     reset:      0          0          0          0                            ok
   touched:   64.2 KiB   31.5 MiB   56.6 GiB  -56.6 GiB                        ok
  segments:     32        504        496          8                            not all freed!
-abandoned:      0          0          0          0                            ok
   -cached:      0          0          0          0                            ok
     pages:      0          0      743.7 Ki  -743.7 Ki                         ok
-abandoned:      0          0          0          0                            ok
 -extended:      0
 -noretire:      0
     mmaps:      0
   commits:   54.4 Ki
   threads:     32         32          0         32                            not all freed!
  searches:     0.0 avg
numa nodes:       1
   elapsed:     239.782 s
   process: user: 2331.109 s, system: 13.125 s, faults: 5442886, rss: 4.8 GiB, commit: 4.9 GiB
```

converted evidenve in 879.csv(Timeline Explorer)
![image](https://user-images.githubusercontent.com/2350416/215297898-b847e433-19de-469d-bd14-aa3399b08e49.png)


### Benchmark2
I ran a benchmark using [hayabusa-sample-evtx](https://github.com/Yamato-Security/hayabusa-sample-evtx) and the results were as follows.

|Version| Elapsed time | Memory(peak)  | Events with hits / Total events | Output file size(bytes)|
|-------| ------------ | --------|--------|--------| 
| [before](https://github.com/Yamato-Security/hayabusa/commit/b97b22403eb4ef4633b7e5a32b499051ae7e395e)  | 00:00:09.073| 1.9 GiB | 19,601 / 76,974 | 16458369  |
| [This PR](https://github.com/Yamato-Security/hayabusa/pull/896/commits/2d89759ab5e56c16a69e6ba7b8fc6641377b5783)   | 00:00:08.639 | 2.0 GiB | 19,601 / 76,974 | 17232165 | 

I would appreciate it if you could review

hitenkoku commented:
in https://github.com/Yamato-Security/hayabusa/pull/896/commits/6505d90b9d3c5788a162cbd342c6f6059d9f26ab
 
|Version| Elapsed time | Memory(peak)  | Events with hits / Total events | Output file size(bytes)|
|-------| ------------ | --------|--------|--------| 
| [before](https://github.com/Yamato-Security/hayabusa/commit/b97b22403eb4ef4633b7e5a32b499051ae7e395e)  | 00:03:52.226| 4.8 GiB | 1,593,715 / 4,817,181 | 575078117 |
| [This PR](https://github.com/Yamato-Security/hayabusa/pull/896/commits/2d89759ab5e56c16a69e6ba7b8fc6641377b5783)   |  00:03:59.750| 5.7 GiB | 1,593,715 / 4,817,181 | 614236764 | 
| https://github.com/Yamato-Security/hayabusa/pull/896/commits/6505d90b9d3c5788a162cbd342c6f6059d9f26ab | 00:03:57.280  | 5.7GiB | 1,593,715 / 4,817,181 | 614236764 |

hitenkoku commented:
@YamatoSecurity  @fukusuket  moved ASN, City, and Country in details field when json-timeline is specified in https://github.com/Yamato-Security/hayabusa/pull/896/commits/693ee7dc3d5f6e58e4a50ce50ff34fe1d404dcb9 .

fukusuket commented on src/detections/configs.rs:
```diff
@@ -619,6 +658,10 @@ pub struct CsvOutputOption {
     #[arg(short, short = 'J', long = "JSON-input")]
     pub json_input: bool,
 
+    /// Scan JSON-formatted logs instead of .evtx
```
```suggestion
    /// Add log enhancement for IP address (ASN, City and Country: Required MaxMind GeoIP databases)
```

fukusuket replied:
I would appreciate it if you could fix the typo :)

hitenkoku replied:
thanks for your review.
i merged your suggestion.

YamatoSecurity commented:
@hitenkoku Thank you so much the PR!
On my dataset, I get the following error when scanning:
```
ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
thread 'hayabusa-thread[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
' panicked at 'thread 'called `Option::unwrap()` on a `None` valuehayabusa-thread', ' panicked at 'src\detections\message.rscalled `Option::unwrap()` on a `None` value:', 146src\detections\message.rs::21146
:note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
21
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: JoinError::Panic(Id(1077978), ...)', src\detections\detection.rs:157:41
```

Can you check this?

YamatoSecurity commented:
Humm.. I tried both CSV and JSON timelines but don't see any of the additional MaxMind fields. Can you check that it is working after the latest commits?

hitenkoku commented:
I checked existed SrcASN etc...  in json and csv in latest commit.
Would you share executed command and environment?

> Humm.. I tried both CSV and JSON timelines but don't see any of the additional MaxMind fields. Can you check that it is working after the latest commits?

hitenkoku commented:
@YamatoSecurity I fixed following problem in https://github.com/Yamato-Security/hayabusa/pull/896/commits/1a02aaf508ecfe85c2c4d607aeb08a394b739e60

Would you check it?

> ```
> ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> thread 'hayabusa-thread[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> ' panicked at 'thread 'called `Option::unwrap()` on a `None` valuehayabusa-thread', ' panicked at 'src\detections\message.rscalled `Option::unwrap()` on a `None` value:', 146src\detections\message.rs::21146
> :note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> 21
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: JoinError::Panic(Id(1077978), ...)', src\detections\detection.rs:157:41
> ```
> 
> Can you check this?



> @hitenkoku Thank you so much the PR! On my dataset, I get the following error when scanning:
> 
> ```
> ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> thread 'hayabusa-thread[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
> ' panicked at 'thread 'called `Option::unwrap()` on a `None` valuehayabusa-thread', ' panicked at 'src\detections\message.rscalled `Option::unwrap()` on a `None` value:', 146src\detections\message.rs::21146
> :note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> 21
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: JoinError::Panic(Id(1077978), ...)', src\detections\detection.rs:157:41
> ```
> 
> Can you check this?

YamatoSecurity commented:
@hitenkoku 
Thank you! The scan continues but I get the following 2 types of errors (blank("") and "LOCAL"):
```
Scanning in progress. Please wait.

48 / 7259 [=>---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------] 0.66 % 140m [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
1074 / 7259 [============================>----------------------------------------------------------------------------------------------------------------------------------------------------------------------] 14.80 % 10m [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
1079 / 7259 [============================>----------------------------------------------------------------------------------------------------------------------------------------------------------------------] 14.86 % 10m [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
1081 / 7259 [=============================>---------------------------------------------------------------------------------------------------------------------------------------------------------------------] 14.89 % 11m [ERROR] IoError("Failed Convert IP Address. input: ")
[ERROR] IoError("Failed Convert IP Address. input: ")
1304 / 7259 [===================================>---------------------------------------------------------------------------------------------------------------------------------------------------------------] 17.96 % 11m [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
1601 / 7259 [===========================================>-------------------------------------------------------------------------------------------------------------------------------------------------------] 22.06 % 10m [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
1897 / 7259 [===================================================>------------------------------------------------------------------------------------------------------------------------------------------------] 26.13 % 9m [ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
[ERROR] IoError("Failed Convert IP Address. input: LOCAL")
```
When it is a local address, I want to output "LOCAL". I am not sure what the blank errors are.

YamatoSecurity commented:
@hitenkoku Thank you, the geoip errors are gone now.
A couple things:
1. Can you change `n/a` to `-` for all of the fields without information?
2. When the IP is `127.0.0.1`, can you output `localhost` in the `SrcASN` or `TgtASN` field?
3. When the IP is a private IP address, (`LOCAL` in the geoIP DB ?), can you output `RFC1918` in the `SrcASN` or `TgtASN` field? (Output `-` for `SrcCountry`, `SrcCity`, etc...)

Everything else looks good.

fukusuket commented:
以下を確認いたしました、 LGTMです！！💪

#### help表示
helpメニューにgeo-ipオプションが表示されること
- hayabusa.exe help csv-timeline(json-timeline)
 
#### geo-ipオプションなし
geo-ipなしの場合、修正前後で以下コマンド結果のdiffがないこと
- hayabusa.exe csv-timeline -d .\hayabusa-sample-evtx -o out.csv (-p super-verbose)
- hayabusa.exe json-timeline -d .\hayabusa-sample-evtx -o out.json (-p super-verbose)

#### geo-ipオプションあり
以下のevtxでマッチするIPがある場合(件数少ないですが)、SrcASN/SrcCountry/SrcCity/TgtASN/TgtCountry/TgtCityがcsv/json出力されること
- hayabusa-sample-evtx
- [いつもの(6.1GB evtx)](https://github.com/Yamato-Security/hayabusa/issues/778#issuecomment-1296504766) 

また [maxminddb-rustのsampleコード](https://github.com/oschwald/maxminddb-rust/blob/main/examples/lookup.rs)の結果と↑出力情報一致も確認いたしました。

#### 確認したサンプル
src
```
{
    "Timestamp": "2021-10-08 21:57:06.763 +09:00",
    "Computer": "fs01.offsec.lan",
    "Channel": "Sec",
    "EventID": 4648,
    "Level": "info",
    "RecordID": 2183520,
    "RuleTitle": "Explicit Logon",
    "Details": {
        "TgtUser": "gentilguest",
        "SrcUser": "admmig",
        "SrcIP": "20.188.56.147",
        "Proc": "",
        "TgtSvr": "printnightmare.gentilkiwi.com",
        "SrcASN": "MICROSOFT-CORP-MSN-AS-BLOCK",
        "SrcCountry": "France",
        "SrcCity": "Paris",
        "TgtASN": "n/a",
        "TgtCountry": "n/a",
        "TgtCity": "n/a"
    }
}
```

dst
```
{
    "Timestamp": "2019-05-02 23:48:53.950 +09:00",
    "Computer": "IEWIN7",
    "Channel": "Sysmon",
    "EventID": 3,
    "Level": "low",
    "RecordID": 10272,
    "RuleTitle": "PowerShell Network Connections",
    "Details": {
        "Proto": "tcp",
        "SrcIP": "10.0.2.15",
        "SrcPort": 49178,
        "SrcHost": "IEWIN7.home",
        "TgtIP": "151.101.36.133",
        "TgtPort": 443,
        "TgtHost": "",
        "User": "IEWIN7\\IEUser",
        "Proc": "C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe",
        "PID": 1508,
        "PGUID": "365ABB72-0244-5CCB-0000-00109AE70B00",
        "SrcASN": "n/a",
        "SrcCountry": "n/a",
        "SrcCity": "n/a",
        "TgtASN": "FASTLY",
        "TgtCountry": "Netherlands",
        "TgtCity": "Amsterdam"
    }
}
```

hitenkoku commented:
@YamatoSecurity 
> 1. Can you change `n/a` to `-` for all of the fields without information?

fixed by 5538694 and 562d3f0.

> 2. When the IP is `127.0.0.1`, can you output `localhost` in the `SrcASN` or `TgtASN` field?

fixed by 63763414d4adb44d566b1c79269120fdd8ed9594

> 3. When the IP is a private IP address, (`LOCAL` in the geoIP DB ?), can you output `Private` in the `SrcASN` or `TgtASN` field? (Output `-` for `SrcCountry`, `SrcCity`, etc...)
> 

When you say 'LOCAL' are you referring to the inputs that came up with the following errors?
`LOCAL` is not in the geoIP DB. value in evtx.`LOCAL` is not in the geoIP DB. value in evtx.

`[ERROR] IoError("Failed Convert IP Address. input: LOCAL")`

hitenkoku commented:
@YamatoSecurity fixed following process by 9f26ca0 and 89e8453.

Would you check it?

> When the IP is a private IP address, (LOCAL in the geoIP DB ?), can you output Private in the SrcASN or TgtASN field? (Output - for SrcCountry, SrcCity, etc...)

YamatoSecurity reviewed (approved):
@hitenkoku 対応ありがとうございました！
これで問題無さそうです。こちらでオプション名を`--GeoIP`に変更して、文面を少し修正しました。
ご確認下さい。これで良ければマージをお願いします。
